src/Meziantou.Analyzer/Rules/AvoidComparisonWithBoolLiteralAnalyzer.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Meziantou.Analyzer.Rules
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class AvoidComparisonWithBoolLiteralAnalyzer : DiagnosticAnalyzer
+    {
+        private static readonly DiagnosticDescriptor s_rule = new DiagnosticDescriptor(
+            RuleIdentifiers.AvoidComparisonWithBoolLiteral,
+            title: "Avoid comparison with bool literal",
+            messageFormat: "Avoid comparison with bool literal",
+            RuleCategories.Style,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true,
+            description: "",
+            helpLinkUri: RuleIdentifiers.GetHelpUri(RuleIdentifiers.AvoidComparisonWithBoolLiteral));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterOperationAction(AnalyzeBinaryOperation, OperationKind.Binary);
+        }
+
+        private static void AnalyzeBinaryOperation(OperationAnalysisContext context)
```

Comment:
That much more robust than the first implementation!

I think you can rewrite `binaryOperation.LeftOperand.Kind == OperationKind.Literal` with `ConstantValue`. I think this can also detect constant usage:

````c#
const bool TrueConstant = true;

bool a = false;
_ = a == TrueConstant; // I think it would detect this
````

Here's what I would have written (not tested):

````c#
// Must have 2 operands
if (binaryOperation.LeftOperand is null || binaryOperation.RightOperand is null)
    return;

// Operands must be in the code
if (binaryOperation.LeftOperand.IsImplicit || binaryOperation.RightOperand.IsImplicit)
    return;

// Operands must be boolean
if(!binaryOperation.LeftOperand.Type.IsBoolean() || !binaryOperation.RightOperand.Type.IsBoolean())
    return;

// Report diagnostic if one operand is a constant
if (IsConstantBooleanValue(binaryOperation.LeftOperand) || IsConstantBooleanValue(binaryOperation.RightOperand))
{
    ReportDiagnostic(...);
}

static bool IsConstantBooleanValue(IOperation operation)
{
    return operation.ConstantValue.HasValue && (operation.ConstantValue.Value == true || operation.ConstantValue.Value == false);
}
````

Reply:
Made changes based on your suggestions (but not quite a copy) and renamed files and classes:

BoolLiteral -> BoolConstant